### DISCUSSION TITLE
How to implement soft-delete

### DISCUSSION THREAD

**Discussion by marinasundstrom** [jasontaylordev/CleanArchitecture#404] [Ideas]
I just want to share how I ended up implementing soft-delete. I wanted to keep it simple.

Feedback is appreciated!

### ISoftDelete interface

Define the interface:

```c#
namespace Sales.Domain.Common
{
    public interface ISoftDelete
    {
        DateTime? Deleted { get; set; }

        string? DeletedBy { get; set; }
    }
}
```

Implement the interface:

```c#
public class Order : AuditableEntity, ISoftDelete 
{
    // Omitted
    
    public DateTime? Deleted { get; set; }

     public string? DeletedBy { get; set; }
}
```

### Modify DbContext

Add this to the ```SaveChangesAsync``` method in your DbContext:

```c#
                    case EntityState.Deleted:
                        if (entry.Entity is ISoftDelete softDelete)
                        {
                            softDelete.DeletedBy = _currentUserService.UserId;
                            softDelete.Deleted = _dateTime.Now;

                            entry.State = EntityState.Modified;
                        }
                        break;
```

### Entity Model Configuration

Configure Query Filter:

```
            builder.Entity<Order>().HasQueryFilter(e => e.Deleted == null);
```

**Comment by iayti** [comment]
Thanks for sharing @robertsundstrom 😊 

Have a nice coding.

**Comment by fakhrulhilal** [comment]
Perhaps this can be moved to wiki instead? Issue is mostly used for tracking item that need to be followed up. I'm afraid, people can't find this useful tutorial when this issue closed.

**Comment by jasontaylordev** [comment]
Thanks for sharing, I've enabled discussions so we can move it over to there

**Comment by ToxinSnake** [comment]
I want to add a small remark to this. You **cannot** use a composite primary key and just slap soft-delete on it. I ran into this issue. I still wanted the database table to reflect the uniqueness of the data.

The codesituation is very simple: There are Vehicles and VehicleActions. A Vehicle can have multiple VehicleActions assigned to it but the same action only once. That's why my first instinct was to just havethe VehicleActionAssignments table have a composite key.

This was my not working code:
```CSharp
public class Vehicle : AuditableSoftDeleteEntity
    {
        public int VehicleId { get; set; }
        //Omitted
    }
```
```CSharp
 public class VehicleAction : AuditableSoftDeleteEntity
    {
        public int VehicleActionId { get; set; }
        //Omitted
    }
```
```CSharp
public class VehicleActionAssignment : AuditableSoftDeleteEntity
    {
        public int VehicleId { get; set; }
        public int VehicleActionId { get; set; }
        //Omitted
    }
```
And in OnModelCreating:
```CSharp
modelBuilder.Entity<VehicleActionAssignment>()
      .HasKey(v => new { v.VehicleActionId, v.VehicleId });
```

Whenever a VehicleActionAssignment gets deleted, it disappears from the frontend. It remains in the database of course. When the user now decides to add that same VehicleActionAssignment back again, it will throw an error.

What I did to resolve this issue is the following:

First I had to add a regular primary key to VehicleActionAssignments
```CSharp
public class VehicleActionAssignment : AuditableSoftDeleteEntity
    {
        public int Id { get; set; }
        public int VehicleId { get; set; }
        public int VehicleActionId { get; set; }
        //Omitted
    }
```
I still wanted the database to force uniqueness of VehicleId and VehicleActionId pairs but have as many deleted same pairs as I want. So what you can do is use **HasFilter**.
```CSharp
modelBuilder.Entity<VehicleActionAssignment>()
    .HasIndex(v => new { v.VehicleActionId, v.VehicleId })
    .HasFilter("Deleted IS NULL")
    .IsUnique();
```
It now forces the pairs to be unique **only** when the Deleted field is null.

**Reply by ToxinSnake** [reply]
Read this for additional information: https://docs.microsoft.com/en-us/ef/core/querying/filters#accessing-entity-with-query-filter-using-required-navigation

**Reply by fakhrulhilal** [reply]
I think that's normal for database design. When delete flag is introduced here, the composite keys don't become unique anymore. The common approach is by introducing new stand alone key (`Id`) like you do above, and use `VehicleActionId` and `VehicleId` as just ordinary foreign key.
